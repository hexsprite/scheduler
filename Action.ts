import { WorkSession } from './WorkSession'

type Duration = number

interface IActionSpec {
  title: string
  duration: Duration
}

export class Action {
  public duration: Duration
  public sessions: WorkSession[]
  public title: string

  constructor(spec: IActionSpec) {
    this.title = spec.title
    this.duration = spec.duration
  }

  public rename(title: string): void {
    this.title = title
  }
}
