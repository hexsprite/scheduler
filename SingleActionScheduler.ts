import { Action } from './Action'
import { WorkSession } from './WorkSession'

let sessions: WorkSession[] = []

export class SingleActionScheduler {
  public scheduleAction(action: Action) {
    let start = new Date()
    for (const session of sessions) {
      start = session.end
    }
    const newSession = new WorkSession({
      duration: action.duration,
      note: '',
      start,
      title: action.title
    })
    sessions.push(newSession)
    action.sessions = [newSession]
  }
}

export function resetSessions() {
  sessions = []
}
