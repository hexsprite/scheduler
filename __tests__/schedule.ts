import * as lolex from 'lolex'
import { Action } from '../Action'
import { resetSessions, SingleActionScheduler } from '../SingleActionScheduler'

describe('scheduler', () => {
  let clock: lolex.Clock
  let scheduler: SingleActionScheduler

  beforeEach(() => {
    clock = lolex.install({ now: new Date('1977-03-21T09:00:00.000Z') })
    scheduler = new SingleActionScheduler()
  })

  afterEach(() => {
    resetSessions()
    clock.uninstall()
  })

  it('schedules single action', () => {
    const action = new Action({ title: 'test action', duration: Minutes(30) })
    scheduler.scheduleAction(action)
    expect(action.sessions).toMatchSnapshot()
  })

  it('schedules two actions', () => {
    const actions = [
      new Action({ title: 'test action', duration: Minutes(30) }),
      new Action({ title: 'test action', duration: Minutes(30) })
    ]
    actions.forEach(action => scheduler.scheduleAction(action))
    expect(actions.map(action => action.sessions)).toMatchSnapshot()
  })
})

const Minutes = (n: number) => n * 1000 * 60
