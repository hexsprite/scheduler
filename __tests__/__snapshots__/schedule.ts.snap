// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scheduler schedules single action 1`] = `
Array [
  WorkSession {
    "duration": 1800000,
    "end": 1977-03-21T09:30:00.000Z,
    "start": 1977-03-21T09:00:00.000Z,
  },
]
`;

exports[`scheduler schedules two actions 1`] = `
Array [
  Array [
    WorkSession {
      "duration": 1800000,
      "end": 1977-03-21T09:30:00.000Z,
      "start": 1977-03-21T09:00:00.000Z,
    },
  ],
  Array [
    WorkSession {
      "duration": 1800000,
      "end": 1977-03-21T10:00:00.000Z,
      "start": 1977-03-21T09:30:00.000Z,
    },
  ],
]
`;
